revolution/motor.py

```diff
@@ -124,3 +128,46 @@ def _revolution(self) -> None:
             with self.environment.contexts() as contexts:
                 contexts.motor_revolution_period = revolution_period
                 contexts.motor_speed = motor_speed
+
+    def cruise_control(self) -> None:
+        with self.environment.contexts() as contexts:
+
+            # values from from cruise_control_pi in GEN11_MCMB
+            k_p = 250.0
+            k_i = 0.015
+            k_d = 0.0
+            
+            integralMin = -200.0
+            integralMax = 200.0
+            integrator = 0.0
+
+            derivativeMin = -100.0
+            derivativeMax = 100.0
+            derivative = 0.0
+
+            outMin = -255.0
+            outMax = 255.0
+            output = 0.0
+
+            timeStep_ms = 20
+            error = 0.0
+            prevError = 0.0
+
+            while (contexts.motor_cruise_control_on):
+                error = contexts.motor_cruise_target_speed - contexts.motor_speed
+                integrator +=  k_i * ((error + prevError) / 2) * timeStep_ms
+                derivative = k_d * (error - prevError) / timeStep_ms
+
+                integrator = np.clip(integrator, integralMin, integralMax)
+                derivative = np.clip(derivative, derivativeMin, derivativeMax)
+
+                output = integrator + derivative + k_p * error
+                output = np.clip(output, outMin, outMax)
+
+                output /= outMax # map to [-1, 1]
```

Comment:
I know you just translated the Gen11 code, but you should actively strive to improve it. For instance, what if ``outMin != -outMax``? Then, this code is not sound. Division must be made by first checking for its sign. I suggest...

```Python
if output > 0:
    output /= outMax
elif output < 0:
    output /= abs(outMin)

Reply:
Ignore the solution I gave above.

From a purely mathematical point of view, these operations don't make sense. Output IS EXPECTED to be in [outMin, outMax] not [-1, 1]. You should just scale them accordingly in Lines 169 and 171.